app/src/androidTest/java/com/google/jetpackcamera/ImageCaptureDeviceTest.kt

```diff
@@ -0,0 +1,176 @@
+            activityScenario!!.result.resultCode
+        }.register("key", ActivityResultContracts.TakePicture()) { result ->
+            assert(result)
+            assert(doesImageFileExist(TEST_URI))
+        }.launch(TEST_URI)
+        deleteFilesInDirAfterTimestamp(timeStamp)
+    }
+
+    @Test // TODO(b/319733374): Return bitmap for external mediastore capture without URI
+    fun image_capture_external_no_uri() = runTest {
+        val launchIntent = Intent()
+        launchIntent.setAction(MediaStore.ACTION_IMAGE_CAPTURE)
+        launchIntent.setComponent(
+            ComponentName("com.google.jetpackcamera", "com.google.jetpackcamera.MainActivity")
+        )
+        getTestRegistry {
+            activityScenario = ActivityScenario.launchActivityForResult(launchIntent)
+            uiDevice.wait(
+                Until.findObject(By.res("CaptureButton")),
+                5000
+            )
+            uiDevice.findObject(By.res("CaptureButton")).click()
+            uiDevice.wait(
+                Until.findObject(By.res("ImageCaptureFailureToast")),
+                5000
+            )
+            activityScenario!!.result.resultCode
+        }.register("key", ActivityResultContracts.TakePicture()) { result ->
+            assert(!result)
+        }.launch(null)
+    }
+
+    @Test
+    fun image_capture_external_illegal_uri() = runTest {
+        val inputUri = Uri.parse("asdfasdf")
+        val launchIntent = Intent()
+        launchIntent.setAction(MediaStore.ACTION_IMAGE_CAPTURE)
+        launchIntent.putExtra(MediaStore.EXTRA_OUTPUT, inputUri)
+        launchIntent.setComponent(
+            ComponentName("com.google.jetpackcamera", "com.google.jetpackcamera.MainActivity")
+        )
+        getTestRegistry {
+            activityScenario = ActivityScenario.launchActivityForResult(launchIntent)
+            uiDevice.wait(
+                Until.findObject(By.res("CaptureButton")),
+                5000
+            )
+            uiDevice.findObject(By.res("CaptureButton")).click()
+            uiDevice.wait(
+                Until.findObject(By.res("ImageCaptureFailureToast")),
+                5000
+            )
+            activityScenario!!.result.resultCode
+        }.register("key", ActivityResultContracts.TakePicture()) { result ->
+            assert(!result)
+        }.launch(inputUri)
+    }
+
+    private fun doesImageFileExist(uri: Uri): Boolean {
+        val file = File(uri.path)
+        if (file.exists()) {
+            val mimeType = URLConnection.guessContentTypeFromName(uri.path)
+            return mimeType != null && mimeType.startsWith("image")
+        }
+        return false
+    }
+
+    private fun deleteFilesInDirAfterTimestamp(timeStamp: Long) {
+        for (file in File(DIR_PATH).listFiles()) {
+            if (file.lastModified() > timeStamp) {
+                file.delete()
+            }
+        }
+    }
+
+    private fun getTestRegistry(launch: () -> Int): ActivityResultRegistry {
```

Comment:
If you change the signature to `launch: (Intent) -> Instrumentation.ActivityResult`, then you can make everything go through the contract.

You would need to change the `onLaunch` code below to:
```
override fun <I, O> onLaunch(
   requestCode: Int,
   contract: ActivityResultContract<I, O>,
   input: I,
   options: ActivityOptionsCompat?
   ) {
    val launchIntent = contract.getIntent(ApplicationProvider.getApplicationContext<Context>(), input)
    val result: ActivityResult = launch(launchIntent)
    dispatchResult(requestCode, result.resultCode, result.resultData)
  }
```